src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -534,24 +544,133 @@ void MarkCustomAttributes (ICustomAttributeProvider provider, in DependencyInfo
 				MarkCustomAttribute (ca, reason);
 				MarkSpecialCustomAttributeDependencies (ca, provider);
 			}
+
+			if (!(provider is MethodDefinition || provider is FieldDefinition))
+				return;
+
+			var dynamicDependencies = _context.Annotations.GetLinkerAttributes<DynamicDependency> ((IMemberDefinition) provider);
+
+			foreach (var dynamicDependency in dynamicDependencies)
+				MarkDynamicDependency (dynamicDependency, (MemberReference) provider);
 		}
 
 		protected virtual bool ProcessLinkerSpecialAttribute (CustomAttribute ca, ICustomAttributeProvider provider, in DependencyInfo reason)
 		{
-			if (IsUserDependencyMarker (ca.AttributeType) && provider is MemberReference mr) {
+			var isPreserveDependency = IsUserDependencyMarker (ca.AttributeType);
+			var isDynamicDependency = ca.AttributeType.IsTypeOf<DynamicDependencyAttribute> ();
+
+			if (!((isPreserveDependency || isDynamicDependency) && provider is MemberReference mr))
+				return false;
+
+			if (isPreserveDependency)
 				MarkUserDependency (mr, ca);
 
-				if (_context.KeepDependencyAttributes || Annotations.GetAction (mr.Module.Assembly) != AssemblyAction.Link) {
-					MarkCustomAttribute (ca, reason);
-				} else {
-					// Record the custom attribute so that it has a reason, without actually marking it.
-					Tracer.AddDirectDependency (ca, reason, marked: false);
+			if (_context.KeepDependencyAttributes || Annotations.GetAction (mr.Module.Assembly) != AssemblyAction.Link) {
+				MarkCustomAttribute (ca, reason);
+			} else {
+				// Record the custom attribute so that it has a reason, without actually marking it.
+				Tracer.AddDirectDependency (ca, reason, marked: false);
+			}
+
+			return true;
+		}
+
+		void MarkDynamicDependency (DynamicDependency dynamicDependency, MemberReference context)
+		{
+			Debug.Assert (context is MethodDefinition || context is FieldDefinition);
+			var member = context as IMemberDefinition;
```

Comment:
As it's used on unlikely paths only, move it there

Reply:
It's also used to get the `Module` when the attribute doesn't give an assembly - which is probably the most common path.